File: R/data-dictionary.R
# Read a data dictionary from disk into the normalized representation the
# rest of the package works from. Only the data-dict.yaml format exists
# today; when other formats arrive (a dbt manifest, harvested catalog
# metadata), this is where the format is inferred and dispatched.
data_dictionary <- function(path) {
  rlang::check_installed("yaml")
  rlang::check_string(path, allow_empty = FALSE)
  new_data_dictionary(yaml::read_yaml(path))
}

as_data_dictionary <- function(x, call = rlang::caller_env()) {
  if (is.null(x) || inherits(x, "commons_data_dictionary")) {
    return(x)
  }
  if (is.character(x) && length(x) == 1) {
    return(data_dictionary(x))
  }
  cli::cli_abort(
    "{.arg dictionary} must be a path to a data-dict.yaml file.",
    call = call
  )
}

new_data_dictionary <- function(raw, call = rlang::caller_env()) {
  raw <- raw %||% list()
  structure(
    list(
      name = prose_field(raw$name),
      description = prose_field(raw$description),
      details = prose_field(raw$details),
      tables = normalize_dictionary_tables(raw$tables, call = call),
      relationships = normalize_dictionary_relationships(raw$relationships),
      glossary = normalize_dictionary_glossary(raw$glossary)
    ),
    class = "commons_data_dictionary"
  )
}

normalize_dictionary_tables <- function(tables, call = rlang::caller_env()) {
  tables <- key_by_name(tables, "table", call = call)
  lapply(tables, function(table) {
    table <- as.list(table)
    table$columns <- key_by_name(table$columns, "column", call = call)
    table
  })
}

# data-dict.yaml lists tables and columns as sequences with a `name` field;
# key entries by that name for lookup. A pre-keyed map is accepted as-is.
key_by_name <- function(entries, what, call = rlang::caller_env()) {
  if (length(entries) == 0) {
    return(list())
  }
  if (all(nzchar(rlang::names2(entries)))) {
    return(entries)
  }

  nms <- vapply(
    entries,
    function(entry) {
      name <- if (is.list(entry)) entry[["name"]]
      if (!is.character(name) || length(name) != 1 || !nzchar(name)) {
        cli::cli_abort(
          "Each {what} in a data dictionary needs a {.field name}.",
          call = call
        )
      }
      name
    },
    character(1)
  )
  entries <- lapply(entries, function(entry) {
    entry[["name"]] <- NULL
    entry
  })
  names(entries) <- nms
  entries
}

normalize_dictionary_relationships <- function(relationships) {
  if (length(relationships) == 0) {
    return(list())
  }
  relationships[vapply(relationships, is.list, logical(1))]
}

normalize_dictionary_glossary <- function(glossary) {
  glossary <- glossary[nzchar(rlang::names2(glossary))]
  lapply(glossary, prose_field)
}

prose_field <- function(x) {
  if (is.null(x)) {
    return(NULL)
  }
  paste(as.character(x), collapse = "\n")
}

# ---- first-touch rendering -------------------------------------------------

# The full entry for one table, delivered when a SQL query touches a table
# the conversation hasn't seen. describe_table composes its own body from
# dictionary_entry_parts() so it can merge the live schema and keep sample
# rows.
dictionary_entry_text <- function(dictionary, table) {
  entry <- dictionary$tables[[table]]
  if (is.null(entry)) {
    return(NULL)
  }
  columns <- dictionary_columns_text(entry$columns)
  if (!is.null(columns)) {
    columns <- paste0("Documented columns:\n\n", columns)
  }
  parts <- c(
    sprintf("Dictionary entry for `%s`:", table),
    dictionary_entry_parts(dictionary, table, columns)
  )
  paste(parts, collapse = "\n\n")
}

dictionary_entry_parts <- function(dictionary, table, columns_text) {
  entry <- dictionary$tables[[table]]
  parts <- c(
    entry$description,
    entry$details,
    columns_text,
    dictionary_relationships_text(dictionary, table)
  )
  c(parts, dictionary_terms_text(dictionary, paste(parts, collapse = "\n")))
}

# One bullet per column. With a live schema (from describe_table), all live
# columns appear, documented ones annotated and undocumented ones keeping
# their inferred type; documented columns missing from the table are flagged.
# Without one (the run_sql path), only documented columns appear.
dictionary_columns_text <- function(columns, live = NULL) {
  if (is.null(live)) {
    if (length(columns) == 0) {
      return(NULL)
    }
    lines <- vapply(
      names(columns),
      function(nm) dictionary_column_line(nm, columns[[nm]]),
      character(1)
    )
    return(paste(lines, collapse = "\n"))
  }

  lines <- vapply(
    seq_len(nrow(live)),
    function(i) {
      dictionary_column_line(
        live$column[[i]],
        columns[[live$column[[i]]]],
        live_type = live$type[[i]]
      )
    },
    character(1)
  )
  out <- paste(lines, collapse = "\n")

  missing <- setdiff(names(columns), live$column)
  if (length(missing)) {
    out <- paste0(
      out,
      sprintf(
        "\n\nDocumented in the dictionary but not present in the table: %s.",
        paste(missing, collapse = ", ")
      )
    )
  }
  out
}

dictionary_column_line <- function(name, spec, live_type = NULL) {
  spec <- spec %||% list()
  qualifier <- paste(
    c(spec$type %||% live_type, spec$units, unlist(spec$constraints)),
    collapse = ", "
  )
  facts <- c(
    spec$description,
    dictionary_values_fact(spec$values),
    dictionary_range_fact(spec$range),
    dictionary_examples_fact(spec$examples),
    spec$details
  )
  detail <- flatten_inline(paste(facts, collapse = " "))

  line <- sprintf("- %s", name)
  if (nzchar(qualifier)) {
    line <- sprintf("%s (%s)", line, qualifier)
  }
  if (nzchar(detail)) {
    line <- sprintf("%s: %s", line, detail)
  }
  line
}

# `values` can be a sequence ([M, F]) or a map of value to meaning
# ({M: Male}).
dictionary_values_fact <- function(values) {
  if (is.null(values)) {
    return(NULL)
  }
  values <- unlist(values)
  labels <- rlang::names2(values)
  rendered <- ifelse(
    nzchar(labels),
    sprintf("%s (%s)", labels, as.character(values)),
    as.character(values)
  )
  sprintf("Values: %s.", paste(rendered, collapse = ", "))
}

dictionary_range_fact <- function(range) {
  if (length(range) < 2) {
    return(NULL)
  }
  sprintf("Range: %s to %s.", as.character(range[[1]]), as.character(range[[2]]))
}

dictionary_examples_fact <- function(examples) {
  if (is.null(examples)) {
    return(NULL)
  }
  sprintf("Examples: %s.", paste(as.character(unlist(examples)), collapse = ", "))
}

dictionary_relationships_text <- function(dictionary, table) {
  relationships <- dictionary$relationships
  mentions <- vapply(
    relationships,
    function(rel) {
      text <- paste(c(rel$join, rel$description), collapse = " ")
      grepl(word_pattern(table), text, ignore.case = TRUE)
    },
    logical(1)
  )
  if (!any(mentions)) {
    return(NULL)
  }

  lines <- vapply(
    relationships[mentions],
    function(rel) {
      head <- paste(
        c(rel$join, if (!is.null(rel$cardinality)) sprintf("(%s)", rel$cardinality)),
        collapse = " "
      )
      desc <- flatten_inline(rel$description %||% "")
      if (!nzchar(head)) {
        return(paste0("- ", desc))
      }
      if (nzchar(desc)) {
        return(paste0("- ", head, ": ", desc))
      }
      paste0("- ", head)
    },
    character(1)
  )
  paste0("Relationships:\n\n", paste(lines, collapse = "\n"))
}

# Definitions of glossary terms the entry references but the system prompt
# doesn't already carry (i.e. terms past the ambient cap).
dictionary_terms_text <- function(dictionary, text) {
  glossary <- dictionary$glossary
  candidates <- setdiff(names(glossary), glossary_ambient(dictionary))
  hits <- candidates[vapply(
    candidates,
    function(term) grepl(word_pattern(term), text, ignore.case = TRUE),
    logical(1)
  )]
  if (length(hits) == 0) {
    return(NULL)
  }
  lines <- sprintf("- %s: %s", hits, flatten_inline(unlist(glossary[hits])))
  paste0("Definitions:\n\n", paste(lines, collapse = "\n"))
}

# Glossary entries are ambient in the system prompt up to a size cap, in
# order of appearance; entries past it are co-resolved at first touch and
# searchable via the context layer.
glossary_ambient <- function(dictionary, cap_chars = 4000) {
  glossary <- dictionary$glossary
  if (length(glossary) == 0) {
    return(character(0))
  }
  sizes <- cumsum(nchar(names(glossary)) + nchar(unlist(glossary)))
  names(glossary)[sizes <= cap_chars]
}

File: pkg-py/src/commons/_data_dictionary.py
"""Read a data dictionary and render the three channels it feeds.

The input is the authored ``data-dict.yaml``, never data-dict's JSON export,
so the schema validated here is the published data-dict spec. Parsing is
permissive about fields commons never inspects and strict about the shape it
does: a ``number(quantity)`` column with no ``range`` reads without complaint,
because nothing reads that range.

Prose fields stay as authored markdown, because they reach the model verbatim.

Reading a dictionary also type-checks any ``definitions:`` blocks against
data-dict's expression language, so an unusable definition raises here,
before any source exists. Only the lowering to SQL waits for a dialect.

The three channels are methods rather than separate structures.
``pkg-r`` spreads the same rendering across ``R/data-dictionary.R``,
``R/prompt.R`` and ``R/context-layer.R``; here the dictionary owns it and the
prompt and context layers call in. That is a difference in shape, not in what
either produces.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any

import yaml
from pydantic import BaseModel, ConfigDict, model_validator

__all__ = [
    "Column",
    "DataDictionary",
    "Definition",
    "Relationship",
    "Table",
    "as_data_dictionary",
]

AMBIENT_GLOSSARY_CAP_CHARS = 4000


def _prose(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, list):
        return "\n".join(str(item) for item in value)
    return str(value)


def _key_by_name(entries: Any, what: str) -> dict[str, Any]:
    """data-dict lists tables and columns as sequences with a `name` field.

    Key them by that name for lookup. A pre-keyed mapping is accepted as-is.
    """
    if not entries:
        return {}
    if isinstance(entries, dict):
        return {str(key): value or {} for key, value in entries.items()}

    keyed: dict[str, Any] = {}
    for entry in entries:
        name = entry.get("name") if isinstance(entry, dict) else None
        if not isinstance(name, str) or not name:
            raise ValueError(f"Each {what} in a data dictionary needs a name.")
        keyed[name] = {key: value for key, value in entry.items() if key != "name"}
    return keyed


class _Permissive(BaseModel):
    model_config = ConfigDict(extra="ignore")


class Column(_Permissive):
    type: str | None = None
    description: str | None = None
    details: str | None = None
    units: str | None = None
    nullable: bool | None = None
    values: Any = None
    range: list[Any] | None = None
    examples: list[Any] | None = None
    constraints: list[str] | None = None


class Definition(_Permissive):
    """One authored governed definition, as written in the YAML.

    `expr` is optional here and required by the export spec, which is what
    type-checks it and reports where it went wrong. Requiring it twice would
    mean two places to keep in step.
    """

    expr: str | None = None
    label: str | None = None
    description: str | None = None
    details: str | None = None
    todo: str | None = None


class Table(_Permissive):
    description: str | None = None
    details: str | None = None
    columns: dict[str, Column] = {}
    definitions: dict[str, Definition] = {}
    # Attached by _definitions at data-source construction; empty until then,
    # so the registry can be exercised without the compiler. Elements are
    # ExportRecord, typed Any so this model need not import _definitions.
    compiled_definitions: list[Any] = []
    # The name the author wrote, kept when a catalog import re-keys this
    # table to the warehouse label, so first touch and relationship matching
    # can still find it.
    authored_name: str | None = None

    @model_validator(mode="before")
    @classmethod
    def _key_children(cls, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        data = dict(data)
        data["columns"] = _key_by_name(data.get("columns"), "column")
        data["definitions"] = _key_by_name(data.get("definitions"), "definition")
        return data


class Relationship(_Permissive):
    join: str | None = None
    cardinality: str | None = None
    description: str | None = None


class DataDictionary(_Permissive):
    name: str | None = None
    description: str | None = None
    details: str | None = None
    tables: dict[str, Table] = {}
    relationships: list[Relationship] = []
    glossary: dict[str, str] = {}
    # Phase 1 of the compiler, keyed by table name. Source-independent, so it
    # is produced here; the SQL needs a dialect and waits for `data_source()`.
    # Values map each definition name to its
    # `_definitions._export.DefinitionExport`, typed loosely so this module
    # need not import the compiler's types.
    definition_exports: dict[str, Any] = {}

    @model_validator(mode="before")
    @classmethod
    def _normalize(cls, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        data = dict(data)
        data["tables"] = _key_by_name(data.get("tables"), "table")
        # Type-checked here rather than at `data_source()` so an unusable
        # definition is reported when the dictionary is read, before any
        # source exists. Only the lowering to SQL needs a dialect.
        from ._definitions._export import export_spec

        data["definition_exports"] = {
            name: table.definitions for name, table in export_spec(data).items()
        }
        for field in ("name", "description", "details"):
            data[field] = _prose(data.get(field))
        relationships = data.get("relationships") or []
        data["relationships"] = [
            item for item in relationships if isinstance(item, dict)
        ]
        glossary = data.get("glossary") or {}
        data["glossary"] = {
            str(term): _prose(body) or "" for term, body in glossary.items() if term
        }
        return data

    @classmethod
    def from_path(cls, path: str | Path) -> DataDictionary:
        """Read a data-dict.yaml file. The YAML is the only input."""
        text = Path(path).read_text(encoding="utf-8")
        return cls.model_validate(yaml.safe_load(text) or {})

    # ---- channel 1: always in the prompt ---------------------------------

    def ambient_prompt_text(self) -> str:
        """The prose that goes into every system prompt for this source."""
        return "\n\n".join(part for part in (self.description, self.details) if part)

    def ambient_glossary_terms(
        self, cap_chars: int = AMBIENT_GLOSSARY_CAP_CHARS
    ) -> list[str]:
        """Glossary terms that fit the ambient cap, in order of appearance.

        Entries past the cap are co-resolved at first touch and searchable
        through the context layer, so nothing is lost by capping.
        """
        terms: list[str] = []
        used = 0
        for term, body in self.glossary.items():
            used += len(term) + len(body)
            if used > cap_chars:
                break
            terms.append(term)
        return terms

    # ---- channel 2: first touch ------------------------------------------

    def entry_text(
        self, table: str, ambient_cap_chars: int = AMBIENT_GLOSSARY_CAP_CHARS
    ) -> str | None:
        """The full entry for one table, delivered the first time it is used."""
        entry = self.tables.get(table)
        if entry is None:
            return None
        columns = self._columns_text(entry)
        if columns is not None:
            columns = f"Documented columns:\n\n{columns}"
        parts = self.entry_parts(table, columns, ambient_cap_chars=ambient_cap_chars)
        return "\n\n".join([f"Dictionary entry for `{table}`:", *parts])

    def entry_parts(
        self,
        table: str,
        columns_text: str | None,
        ambient_cap_chars: int = AMBIENT_GLOSSARY_CAP_CHARS,
    ) -> list[str]:
        """The body of a table's entry, without the heading.

        `describe_table` composes its own body from these, so it can merge the
        live schema and keep sample rows.
        """
        entry = self.tables.get(table)
        if entry is None:
            return []
        # Imported here because _definitions does not import this module and
        # this keeps it that way.
        from ._definitions import entry_text as definitions_entry_text

        parts = [
            part
            for part in (
                entry.description,
                entry.details,
                columns_text,
                definitions_entry_text(entry.compiled_definitions),
                self._relationships_text(table, entry.authored_name),
            )
            if part
        ]
        terms = self._terms_text("\n".join(parts), ambient_cap_chars)
        return [*parts, terms] if terms else parts

    def _columns_text(self, entry: Table) -> str | None:
        if not entry.columns:
            return None
        return "\n".join(
            _column_line(name, column) for name, column in entry.columns.items()
        )

    def _relationships_text(
        self, table: str, authored_name: str | None = None
    ) -> str | None:
        """Relationships mentioning this table, under either of its names.

        A catalog import re-keys a table to its warehouse label, while the
        relationship prose still says what the author wrote, so both names
        have to match or the join disappears from the entry.
        """
        names = [table, authored_name] if authored_name else [table]
        lines = []
        for relationship in self.relationships:
            text = " ".join(
                part for part in (relationship.join, relationship.description) if part
            )
            if not any(_word_pattern(name).search(text) for name in names):
                continue
            head = " ".join(
                part
                for part in (
                    relationship.join,
                    f"({relationship.cardinality})"
                    if relationship.cardinality
                    else None,
                )
                if part
            )
            body = _flatten_inline(relationship.description or "")
            if not head:
                lines.append(f"- {body}")
            elif body:
                lines.append(f"- {head}: {body}")
            else:
                lines.append(f"- {head}")
        if not lines:
            return None
        return "Relationships:\n\n" + "\n".join(lines)

    def _terms_text(self, text: str, ambient_cap_chars: int) -> str | None:
        """Glossary terms the entry mentions that the prompt does not carry."""
        ambient = set(self.ambient_glossary_terms(ambient_cap_chars))
        hits = [
            term
            for term in self.glossary
            if term not in ambient and _word_pattern(term).search(text)
        ]
        if not hits:
            return None
        lines = [f"- {term}: {_flatten_inline(self.glossary[term])}" for term in hits]
        return "Definitions:\n\n" + "\n".join(lines)

    # ---- channel 3: the search index -------------------------------------

    def context_chunks(self) -> list[str]:
        """Table- and dictionary-level prose, chunked for retrieval.

        Column-level content stays out: first touch owns it, and indexing it
        would pay for a second copy the agent already has.
        """
        chunks: list[str] = []
        if self.details:
            chunks.append(self.details)
        for name, entry in self.tables.items():
            prose = "\n\n".join(
                part for part in (entry.description, entry.details) if part
            )
            if prose:
                chunks.append(f"Table `{name}`: {prose}")
        chunks.extend(f"{term}: {body}" for term, body in self.glossary.items())
        from ._definitions import context_chunks as definitions_context_chunks

        for entry in self.tables.values():
            chunks.extend(definitions_context_chunks(entry.compiled_definitions))
        return [chunk for chunk in chunks if chunk]


def as_data_dictionary(x: Any) -> DataDictionary | None:
    if x is None or isinstance(x, DataDictionary):
        return x
    if isinstance(x, (str, Path)):
        return DataDictionary.from_path(x)
    raise TypeError(
        f"dictionary must be a path to a data-dict.yaml file, got {type(x).__name__}."
    )


def _flatten_inline(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def _word_pattern(word: str) -> re.Pattern[str]:
    return re.compile(rf"(?<!\w){re.escape(word)}(?!\w)", re.IGNORECASE)


def _column_line(name: str, spec: Column) -> str:
    qualifier = ", ".join(
        str(part)
        for part in (
            spec.type,
            _nullability_fact(spec.nullable),
            spec.units,
            *(spec.constraints or []),
        )
        if part
    )
    facts = [
        part
        for part in (
            spec.description,
            _values_fact(spec.values),
            _range_fact(spec.range),
            _examples_fact(spec.examples),
            spec.details,
        )
        if part
    ]
    detail = _flatten_inline(" ".join(facts))

    line = f"- {name}"
    if qualifier:
        line = f"{line} ({qualifier})"
    if detail:
        line = f"{line}: {detail}"
    return line


def _nullability_fact(nullable: bool | None) -> str | None:
    if nullable is None:
        return None
    return "nullable" if nullable else "not nullable"


def _values_fact(values: Any) -> str | None:
    """`values` is a sequence ([M, F]) or a map of value to meaning ({M: Male})."""
    if not values:
        return None
    if isinstance(values, dict):
        rendered = [f"{value} ({label})" for value, label in values.items()]
    else:
        rendered = [str(value) for value in values]
    return f"Values: {', '.join(rendered)}."


def _range_fact(bounds: list[Any] | None) -> str | None:
    if not bounds or len(bounds) < 2:
        return None
    return f"Range: {bounds[0]} to {bounds[1]}."


def _examples_fact(examples: list[Any] | None) -> str | None:
    if not examples:
        return None
    return f"Examples: {', '.join(str(item) for item in examples)}."
